### ISSUE TITLE
Edit course page

### ISSUE THREAD

**Issue by qurat08** [ude-soco/CourseMapper-webserver#1258]
User lands on course description page in two ways:
1) clicking on the course card on home or landing page
2) clicking on course initials in the square button on the left nav bar

The layout of the course description page should be changed to: (If you wish to see it in figma, its here: https://www.figma.com/design/Iy6e4BqclQh17KDU78ihI4/CourseMapper?node-id=2001-24071&node-type=canvas&t=Pg6Q7oUcR6oo9maI-0)

before enrolling in the course:
![image](https://github.com/user-attachments/assets/26b60094-233b-42fe-8427-05c5f9cc8868)

after enrolling in the course:
![image](https://github.com/user-attachments/assets/688ef755-f0c2-4196-a81d-4d6b40e23185)

for course moderator (Edit buttons)
![image](https://github.com/user-attachments/assets/f2040639-b153-4800-8ec5-0496a0303b8e)

When course moderator clicks on edit icon close to course name, rename functionality should be applied as it is for now.

When course moderator clicks on edit course image icon:
this is the complete issue: https://github.com/ude-soco/CourseMapper-webserver/issues/1107

The same image will appear on the course card on home and landing page

When course moderators clicks on edit description icon, the text formatting options should appear (one option is to use the p-editor from PimeNG https://primeng.org/editor)
![image](https://github.com/user-attachments/assets/5394f653-e47e-43af-8e94-390d7d88a56e)

Once anything (course name, description, display image) is edited, the save and cancel button appears at the bottom and clicking on save button saves the changes and show the success message as needed i.e. 
Course display picture is updated
Course name is updated
Course description is updated.

The moderator should be able to land on the course description page by using a new option 'Edit course' in the three dot menu with course name in the left nav bar (with this option the icon should be pencil icon used for editing generally). This three dot menu needs to be updated as: 
![image](https://github.com/user-attachments/assets/0c544e8e-6a27-4042-926b-e65d8008fb9c)
Rename, Edit course, Share course, View course dashboard, Notification settings, Delete course.... the order is changed, just a new option Edit course is added... all these options are shown to the moderator, however, the only ones available to the enrolled user are: Share course, View course dashboard, Notification settings

**Comment by qurat08** [comment]
a new update on https://github.com/ude-soco/CourseMapper-webserver/issues/1107

**Comment by qurat08** [comment]
update: Please include the number of enrolled users too: 
![image](https://github.com/user-attachments/assets/83fc2d6d-e9c2-4ad5-845d-4366be887672)

**Comment by ahmedmohyh** [comment]
> update: Please include the number of enrolled users too: ![image](https://private-user-images.githubusercontent.com/23054665/390755527-83fc2d6d-e9c2-4ad5-845d-4366be887672.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzM1NzYwNjMsIm5iZiI6MTczMzU3NTc2MywicGF0aCI6Ii8yMzA1NDY2NS8zOTA3NTU1MjctODNmYzJkNmQtZTljMi00YWQ1LTg0NWQtNDM2NmJlODg3NjcyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjA3VDEyNDkyM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTkyYjU1YWU2NjUzNmMwNDRjMzA4YjZlNjI1YTdiNTljODE3ZTMxNmRjYmZmMGJlYWM5MGRjODYwOWYwMmZiMDYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.X6BtpNPP0L3cZfLUTCG-ckG_O0YbaaQowrOzEecfViA)

do we need also the number of the enrolled users here as well?

<img width="299" alt="image" src="https://github.com/user-attachments/assets/e390d24d-4900-4525-8ae6-6b25dc7d5987">

**Comment by qurat08** [comment]
This is already done by hager in her thesis. These changes will be reflected when merged later.

**Comment by qurat08** [comment]
The final UI is different in some aspects than the provided design:
*Before enrolling in*
1. Is it intentionally center aligned? If yes, was it discussed with anyone?
2. Please see the difference in font of 'enrolled users' and 'created by and created on', all of the three should look exactly the same. 
3. Text in the button 'Enroll In Course' should be same as in the prototype
4. There is a lot of extra space between the line and the enroll button, try to use the space equally above and below the button
5. Please check if the font style of the heading 'course description' same as of 'Created by', it looks a bit different

![Image](https://github.com/user-attachments/assets/99637ea1-3768-4655-8546-96886c7288e4)

*after enrolling in*
1. Alignment is a bit off. Everything is required to align left but course name and course description is aligned more towards left than the rest. Either shift these two with created by, created on, enrolled users and enroll button or move the rest to the left. 
2. Moreover, the text with the icons is also not aligned with each other. All the icons and texts should be exactly aligned with one over the other
3. Change the text on the button to 'Unenroll from course' as in the prototype
4. Move the button one line downwards so it appears with a gap from the above text
5. If no description is provided, a text should be shown as 'No description provided' rather than keeping it empty.
6. the picture should take more space as in the design, till the top of the course name.
7. what is this text with enrolled users?

![Image](https://github.com/user-attachments/assets/1ff97ccf-b1ff-4337-9fd9-700a910218ff)

**Comment by qurat08** [comment]
1. The default pictures assigned to courses keep on changing after a few seconds. A fixed random picture should be assigned to each course so that the user can remember the course by the picture as well. Changing them randomly after some time would not work.

2. The course picture is not fitting to the space allocated to ti on the card, therefore wasting soem space on the card as well as disturbing the size of the cards (some become longer than others). The picture should fill in the whole space assigned to it and the size of the card should not be disturbed 

![Image](https://github.com/user-attachments/assets/26ebf31f-c6c5-4fe0-892f-7716942dad52)

3. On the course card on home and landing page, number of enrolled users should be added. This is done in Hager's branch but since we will not deploy it now, this information needs to be added explicitly on the course card.

![Image](https://github.com/user-attachments/assets/63e4adcd-f16c-4609-a4a4-96ad58d857b4)

4. Is it possible to move the edit pictúre icon on the image? Because if its under the picture it takes some space from the bottom. The icon should be there all the time on the top right cirner of the image and visible to the user.

![Image](https://github.com/user-attachments/assets/c2ed1d67-970f-4332-8a86-e8d1a0c087f2)
To make it more visible a grey circle can be added to the edit button so its visible all the time no matter what the color of the picture is

5. Issue with edit description panel: 
There is a lot of space on th epage but edit description panel is too small. It should grow till the maximum space on the page whiel the user is writing in it. Once th emax space is reached, it should become scrollable.
now: 

![Image](https://github.com/user-attachments/assets/a2c2dc05-2991-4877-90ec-6875dac64e07)
expected:

![Image](https://github.com/user-attachments/assets/795b35d3-e7bd-4e85-8c8c-77797e7fc11e)

6. If I try to upload an image in course description, it does not appear. If this is not possible, then we do not provide this option in editabel panel and keep it simple

![Image](https://github.com/user-attachments/assets/387a19e2-9768-4f63-83f7-62326b1304c1)

7. Save and cancel button should be with a tick and cross icon with the text and their position should be swapped as in the prototype
now: 

expected: this is how we have tick and cross icon on the buttons with text in course creation:
![Image](https://github.com/user-attachments/assets/8cc23ea5-9b1e-4d62-9da7-113e66e95ed2)

prototype:
![Image](https://github.com/user-attachments/assets/4e14992f-f554-4fb2-bb11-1aa65cfd3b65)

**Comment by qurat08** [comment]
update: i just noticed that showing enrolled users on cards on home and landing page is already done

**Comment by qurat08** [comment]
the random images that should be used on coursecards can be found here: SoCo_New\Teaching\Theses\Project\CourseMapper\Random images\Final

**Comment by rawaa123** [comment]
> 1. The default pictures assigned to courses keep on changing after a few seconds. A fixed random picture should be assigned to each course so that the user can remember the course by the picture as well. Changing them randomly after some time would not work.
>    I will look for some random poictures and provide to you

This is done for the newely created course but for the exist ones will assign one random img. I cant upload all imgs to frontend because this will harm the performance & bundle size, due to Including 60 MB of images in frontend bundle would dramatically increase the initial load time. Therfore, I uploaded them all to the backend. Also I cant assign random imgs to the already created courses as i have no idea which one assigned an img by the moderator and which one has a default img as all of them have this url value.

**Comment by qurat08** [comment]
Is it possible that you leave the existing ones as it is, meaning that with soco image.

Just think about the deployed version, we have our labs and courses there with soco image. If you change them (since they are already created courses) with any random images, it will not look good. 

So if in future, a course is created without an image, we assign it an image from the folder I shared in sciebo. The same happens if someone removes their course image.
For existing courses, we leave them as it is with soco image.

Is that possible? What do you think of this solution?

**Comment by rawaa123** [comment]
> 1.1 When edit icon on image is clicked, the file selector should open directly and the image should be uploaded and preview is shown in the same space, with save and cancel buttons enabled. If the user click on save, the image is saved, otherwise reverted back to the previous version. (This is the current version of the prototype)
> 
> 1.2 If you want to go with the current implementation, then the size of the box showing uploaded image etc, need to be adjusted with the image size etc.. currently it grows to the left and looks weird. if we have to keep this the size should not grow everytime like this

1.1 This is the best thing I can do, once the user clickes on the edit icon then will open the selectin dialog directly based on the current version of primng that we have. 

![Image](https://github.com/user-attachments/assets/afd16d21-bb2f-478f-a5b6-1beb40a3a6c6)

1.2. this will prevent the thumb and file name to expand and ensures that both the width and height of the file upload area remain constrained. In any case, if you want to present the thumb and name of he file then it must expand to show both. If you dont want it to expand at all then i have to remove the thumb.  

![Image](https://github.com/user-attachments/assets/55d421c7-6980-4843-ac6e-d40e1b53652b)

> 3\. On the course card on home and landing page, number of enrolled users should be added. This is done in Hager's branch but since we will not deploy it now, this information needs to be added explicitly on the course card.

This is already implemented and in both pages enrolled users is implemnetd. It is available in three different places: Lnading page, home page and welcome page. It is not cleare to me where u want to add it

**Comment by qurat08** [comment]
> 1.1 This is the best thing I can do, once the user clickes on the edit icon then will open the selectin dialog directly based on the current version of primng that we have

If thats possible, please go for this. Showing the file name does not add anything and also not look good. 
When clicking on edit, file selector opens and the image is uploaded and shown in the box (where the previous image was), when clicking on save button in the end, image is uploaded. If save not clicked, its discarded.



> This is already implemented and in both pages enrolled users is implemnetd. It is available in three different places: Lnading page, home page and welcome page. It is not cleare to me where u want to add it

Yes, I noticed that and commented that on another issue if not this one. Thats sorted. Thanks.

**Comment by rawaa123** [comment]
> If thats possible, please go for this. Showing the file name does not add anything and also not look good.
> When clicking on edit, file selector opens and the image is uploaded and shown in the box (where the previous image was), when clicking on save button in the end, image is uploaded. If save not clicked, its discarded.

what do u mean "where the previous image was"???
if you mean about the thumb then this is doable, if u mean to place instead of the original one, then this is not applicable by the primeng. 
Please, refer to primeng to see posible options offered.

**Comment by qurat08** [comment]
> 1.1 This is the best thing I can do, once the user clickes on the edit icon then will open the selectin dialog directly based on the current version of primng that we have.

I understood by this that we will have no button to choose file but just edit button will open file selector. Is that correct or clicking on choose file button will open file selector?



> if you mean about the thumb then this is doable, if u mean to place instead of the original one, then this is not applicable by the primeng.

I meant it instead of the previous one (but I got that its not possible so we ignore it). Showing the image on thumb would look even worse than showing file name on the thumb. 



> If you dont want it to expand at all then i have to remove the thumb.

In this case, what would happen after I select a file?

One possible solution could be to use the same file selector that we use to upload a pdf, the selected pdf name appears on the button and no other area is expanded. While there is no image selected, choose file button appears as it is, without 'no file chosen text' and once a file is uploaded the name appears on the button (button expansion should not go beyong the width of the image on top to maintain the space), in this way we will avoid the expansion that happens under the button. What do you think about this?

![Image](https://github.com/user-attachments/assets/80285f40-9860-4467-9284-39250e79295f)

**Comment by rawaa123** [comment]
> One possible solution could be to use the same file selector that we use to upload a pdf, the selected pdf name appears on the button and no other area is expanded. While there is no image selected, choose file button appears as it is, without 'no file chosen text' and once a file is uploaded the name appears on the button (button expansion should not go beyong the width of the image on top to maintain the space), in this way we will avoid the expansion that happens under the button. What do you think about this?

I followed the following scenario:
once the edit icon is cliked then this will open the file dialog to select the img as shown in the screenshot. 

![Image](https://github.com/user-attachments/assets/3866dee0-4a15-49bf-89db-0c4ed9e79727)

and once the file is selected then the name appears on the button as shown in the screenshot, and no expansion

![Image](https://github.com/user-attachments/assets/16e8496d-441b-41fd-bc01-e64f25535f60)

**Comment by rawaa123** [comment]
There is a problem in the code that the one who implemeted this is using the same save and cancel buttons for both edit panels image and description, along with some inconsistancy in the code. I will not work on this right now to separate them and update some features, will leave as a future work

**Comment by qurat08** [comment]
> the success message as needed i.e.
> Course display picture is updated
> Course name is updated
> Course description is updated.

1. These success messages are not shown. I have no indication after saving, if my changes are saved or not becasue I dont see these success messages.

2. When I delete course image, its not working. 

3. Clicking on edit course in this three dot menu, opens the file selector to upload image. The fiel selector to uplaod image should only open when we are trying to edit course image, not on edit course. May be opening file selector by clicking on 'choose a file button' will help fixing this.

![Image](https://github.com/user-attachments/assets/060c4afd-e09c-4e72-ae7f-4e94c988535f)

> once the edit icon is cliked then this will open the file dialog to select the img as shown in the screenshot.

in this case, the 'choose file' button seems to be useless. If we have this button, then file selector should open by clicking on this button. This will introduce more number of clicks, first on edit icon, then on choose file,which is not optimum, but in the current scenario, choose file button seem to look useless.

---
I am not sure if these issues still exist or not because I faced them while there was an issue with random inages folder on dev... (discussed with Rawaa on teams) tillt hat issue is not fixed, I will keep this comemnt and test this again once its fixed and will update.

**Comment by rawaa123** [comment]
I havent worked on 1 and 3 as i have been instructed to start doing things commnunicated on [on Jan 23](https://github.com/ude-soco/CourseMapper-webserver/issues/1258#issuecomment-2609875213)


> in this case, the 'choose file' button seems to be useless. If we have this button, then file selector should open by clicking on this button. This will introduce more number of clicks, first on edit icon, then on choose file,which is not optimum, but in the current scenario, choose file button seem to look useless.

No, the file selector will open directly after clicking on edit icon as i have been instructed to do above, but if the user wants to change the picture then he has to click on the button. 


> I am not sure if these issues still exist or not because I faced them while there was an issue with random inages folder on dev... (discussed with Rawaa on teams) tillt hat issue is not fixed, I will keep this comemnt and test this again once its fixed and will update.

This because the one who worked was storing imgaes in PDF file which is incorrect. we need to store them in a separete folder named imgaes and this file is not exist in the server. This cuses the problem. 
I checked the server side two folders are missing and reported this to Ralf in this [issue ](https://github.com/ude-soco/CourseMapper-webserver/issues/1498), so he can pull the new changes

**Comment by qurat08** [comment]
> No, the file selector will open directly after clicking on edit icon as i have been instructed to do above,

You were only instructed to do so based on the possibilities you proposed and based on what was initially implemented by Mousa. In the initial design this button never existed or was needed. But how he implemented introduced this button. We can just keep it like this and can refer to this discussion if anything needs to be improved by someone in future.

**Comment by qurat08** [comment]
> 1. These success messages are not shown. I have no indication after saving, if my changes are saved or not becasue I dont see these success messages.

Update: these messages are still missing. May be someone can work on them in future.

**Comment by rawaa123** [comment]
> You were only instructed to do so based on the possibilities you proposed and based on what was initially implemented by Mousa. In the initial design this button never existed or was needed. But how he implemented introduced this button. We can just keep it like this and can refer to this discussion if anything needs to be improved by someone in future.

This is offered by Primeng. You may check possible opetions offered.

**Comment by qurat08** [comment]
> > You were only instructed to do so based on the possibilities you proposed and based on what was initially implemented by Mousa. In the initial design this button never existed or was needed. But how he implemented introduced this button. We can just keep it like this and can refer to this discussion if anything needs to be improved by someone in future.
> 
> This is offered by Primeng and the avilble version of it. You may check possible opetions offered.

Yes, this is what you mentioned earlier and we already discussed and agreed upon.

**Comment by rawaa123** [comment]
> 1. These success messages are not shown. I have no indication after saving, if my changes are saved or not becasue I dont see these success messages.

This i can be done this release not an issue.

**Comment by qurat08** [comment]
> > 1. These success messages are not shown. I have no indication after saving, if my changes are saved or not becasue I dont see these success messages.
> 
> This i can be done this release not an issue.

Thank you very much.

**Comment by rawaa123** [comment]
> 3\. Clicking on edit course in this three dot menu, opens the file selector to upload image. The fiel selector to uplaod image should only open when we are trying to edit course image, not on edit course. May be opening file selector by clicking on 'choose a file button' will help fixing this.

Sorry, after the discussion, what can we conclude?? 
The user should click on the edit icon to select a picture and this will open the dialog immediately, and if he needs under any circumstances to change the selected one then will clik on the button. To save it, then he clicks save. 
This is the scenario given right now.

**Comment by qurat08** [comment]
> > 3. Clicking on edit course in this three dot menu, opens the file selector to upload image. The fiel selector to uplaod image should only open when we are trying to edit course image, not on edit course. May be opening file selector by clicking on 'choose a file button' will help fixing this.
> 
> Sorry, after the discussion, what can we conclude?? The user should click on the edit icon to select a picture and this will open the dialog immediately, and if he needs under any circumstances to change the selected one then will clik on the button. To save it, then he clicks save. This is the scenario given right now.

Yes. But, this point is related to the three dot menu with the course name on course page. That was opening the file selector... rather than opening edit course page.

**Comment by rawaa123** [comment]
> Yes. But, this point is related to the three dot menu with the course name on course page. That was opening the file selector... rather than opening edit course page.

Will check it to fix it. Thanx for pointing this out

**Comment by rawaa123** [comment]
Clicking on edit course in this three dot menu, what should open exactly, course title or description to be edited.

**Comment by qurat08** [comment]
> Clicking on edit course in this three dot menu, what should open exactly, course title or description to be edited. Currently opens three of them; img, title and description

it should just redirect to the editable course page where he can edit whatever he wants to edit. this page i meanvhttps://br-dev.coursemapper.de/course/xyz/welcome
This will have the editing options and he can edit whatever he wants to. Also becasue edit course option is only shown to the moderator in the three dot menu, so the welcome page he will land on will always be with editing options like currently.

**Comment by qurat08** [comment]
1. create a new course without picture is working, it assigns a random image.
2. create a new course with picture is working
3. Editing (uploading a new image to a course created without picture is workiing)
4. Editing (uploading a new image to a course created with picture is workiing)
5. delete a picture in both courses is also working. After deletion, a random image is assigned as desired.
6. Deleting a picture which was randomly assigned to either a new course created without picture, or after deleting a picture, does not work. (I think it is implamented like this that if a random mage is assigned, its deletion will assign another random image which is actually there so it will not change it. just mentioning it for the record, no change is needed)

**Comment by qurat08** [comment]
> Clicking on edit course in this three dot menu, what should open exactly, course title or description to be edited. Currently opens three of them; img, title and description

The problem I noticed does not happen when I click on edit course option in three dot menu, while I am already in another page e.g. on a material or in a channel etc.
The problem is that if I am already on edit course page (https://brdev.coursemapper.de/course/xyz/welcome) and I click on edit course option in the three dot menu, file selector opens, ideall nothing should happen as I am already on that page. (This is a rare case which I just pointed out as I noticed, you can decide if any action needs to be taken or leave as it is)

**Comment by qurat08** [comment]
I noticed two new problems on the same page so created two seperate issues for them #1507 #1508

**Comment by qurat08** [comment]
> > 1. These success messages are not shown. I have no indication after saving, if my changes are saved or not becasue I dont see these success messages.
> 
> This i can be done this release not an issue.

once this is done, the issue can be closed

**Comment by rawaa123** [comment]
> 1. These success messages are not shown. I have no indication after saving, if my changes are saved or not becasue I dont see these success messages.

Based on the current implementation and time limit, this should be combined in one msg

Course name is updated
Course description is updated.

I changed this to Course content is updated. Would you s uggest sth else. 
However, in the future return me back here because i need to rectify the code and the UI for save and cancel buttons

**Comment by rawaa123** [comment]
> The problem is that if I am already on edit course page (https://brdev.coursemapper.de/course/xyz/welcome) and I click on edit course option in the three dot menu, file selector opens, ideall nothing should happen as I am already on that page

done

**Comment by qurat08** [comment]
> Would you s uggest sth else.

Course details are updated successfully.

**Comment by rawaa123** [comment]
I found an issue. If I am on edit course page and and update the description of the course. Then land on another page e.g material, channgel and want to update course details throug clicking on edit course option in the three dot menu,  then the old course description will appaer not the new one. 

this is solved as well

**Comment by rawaa123** [comment]
Another issue has been detected: If I am in course edit page and navigate to material or channel  and want to update course details throug clicking on edit course option in the three dot menu, then  the number of enrolled user will be set to 0.

![Image](https://github.com/user-attachments/assets/6a8fa069-866f-40a9-85d1-fe6720964db8)

This is after navigation from material taps, alos solved. 

![Image](https://github.com/user-attachments/assets/bee56076-54c5-4f13-bc0e-a84c30937075)

**Comment by qurat08** [comment]
> I found an issue. If I am on edit course page and and update the description of the course. Then land on another page e.g material, channgel and want to update course details throug clicking on edit course option in the three dot menu, then the old course description will appaer not the new one.

this is not fixed, If I edit something on edit course page and do not click on 'save' and move to another course and click on edit course, the details from the first course are shown here.

![Image](https://github.com/user-attachments/assets/8fdab7f7-014d-4bfd-b8bd-45442498878d)


If I am not a moderator of a course and land on a course while I have unsaved changes from another course description, they appear here on this course which I am not the owner of: 

![Image](https://github.com/user-attachments/assets/3420c577-e2a9-42f6-9406-52fc94f298b1)

![Image](https://github.com/user-attachments/assets/92f8c5c0-7011-47f4-b2de-45dc25730a2c)

Also, while uploading an image for the course card during course creartion, the selected image anme appears twice, once on the button and once under it. 

![Image](https://github.com/user-attachments/assets/0c710155-dafe-462f-8482-e7a5552516bd)

**Comment by qurat08** [comment]
There is another problem that save and cancel buttons appear at the very end of the page, the user is not aware if he has to click on these buttons to save changes or not becasue they are not visible. They should appear on the UI without making it scrollable so the user can visibly notice that teh appear when any edit icon is clicked. 

![Image](https://github.com/user-attachments/assets/0681fde7-7172-4010-977f-69a5864a7257)

**Comment by rawaa123** [comment]
> I found an issue. If I am on edit course page and and update the description of the course. Then land on another page e.g material, channgel and want to update course details throug clicking on edit course option in the three dot menu, then the old course description will appaer not the new one.

This is done if saved is clicked

**Comment by qurat08** [comment]
> This rasises in case saved is not clicked

yes, and the problem is with the save button appearing not visible at the end. if we move save abd cancel a bit up so that they are always visible to the user, then this may trigger them to click on save button. Is that possible?

Did you update anything related to this issue based on our last discussion? because this is still the same and I was wondering if anything is updated or not?

**Comment by rawaa123** [comment]
> yes, and the problem is with the save button appearing not visible at the end. if we move save abd cancel a bit up so that they are always visible to the user, then this may trigger them to click on save button. Is that possible?

I am doing my best to meove it up but dosent look good



> Did you update anything related to this issue based on our last discussion? because this is still the same and I was wondering if anything is updated or not?

Unfortunately, the only thing was able to do is preventing user from navigating to another page if editing without saving but when navigating to another course is still not workable, I am still investigating solutions that aligns with the current implementaion for course selection. You may check the implemantaion of the three relted components. I am open for any suggesiotns or ideas that alignes with currect implementstion.
